### DISCUSSION TITLE
Is there any way to specify the embedding dimensionality via llama.cpp?

### DISCUSSION THREAD

**Discussion by Stevenic** [ggml-org/llama.cpp#6618] [Q&A]
I'm using [nomic-embed-text-v1.5](https://huggingface.co/nomic-ai/nomic-embed-text-v1.5-GGUF) for my embedding model and it works but it's returning the full 768 dimensions. I'd like to drop this to [128 dimensions](https://huggingface.co/nomic-ai/nomic-embed-text-v1.5) but I don't see a way to do that via llama.cpp.

**Comment by ggerganov** [comment]
We can try to add this as an option but need a reference to the algorithm that performs the dimensionality reduction

**Reply by Stevenic** [reply]
The embeddings I'm trying to use are from nomic and the model card is here:
https://huggingface.co/nomic-ai/nomic-embed-text-v1.5

They mention this paper for the dimensionality pieces, which I believe is the same technique that OpenAI is using to achieve variable length embeddings as well:
https://arxiv.org/abs/2205.13147

Their API example points to a python library here:
https://github.com/nomic-ai/nomic/blob/main/nomic/embed.py#L194

But digging into that it looks like it's just calling a service their hosting. I need this to all happen offline so that won't work for me.

Let me research this a bit more...

**Reply by Stevenic** [reply]
@cebtenzzre do you know if there's a way to get only a 128 dimension nomic embedding via llama.cpp instead of the 768 dimension embedding that the provided GGUF results in?

**Reply by Stevenic** [reply]
BTW... Love what you folks are doing here @ggerganov ... I just ordered a new machine with dual RTX 4090's and I'm fighting the urge to dust off my C++ chops and dig into everything you folks are building :)

**Comment by cebtenzzre** [comment]
This is how the dimensionality is resized using Sentence Transformers or HF Transformers: https://huggingface.co/nomic-ai/nomic-embed-text-v1.5#sentence-transformers

I wrote a C++ implementation of this for GPT4All (which is built on llama.cpp), the important part is here: https://github.com/nomic-ai/gpt4all/blob/ac498f79ac67640281bbf85137a5c632c59400ef/gpt4all-backend/llamamodel.cpp#L871-L885

The easiest way to leverage that implementation is to `pip install "nomic[local]"` and then follow [this example](https://docs.nomic.ai/atlas/guides/embeddings#resizable-embeddings) but add `inference_mode="local"`, which causes the nomic client to use GPT4All to generate embeddings instead of the Atlas API. Documentation for that parameter will be added soon.

**Comment by hasaranga** [comment]
@ggerganov 

```
def normalize_l2(x):
    x = np.array(x)
    if x.ndim == 1:
        norm = np.linalg.norm(x)
        if norm == 0:
            return x
        return x / norm
    else:
        norm = np.linalg.norm(x, 2, axis=1, keepdims=True)
        return np.where(norm == 0, x, x / norm)
```

ref: [https://huggingface.co/nomic-ai/nomic-embed-text-v1.5-GGUF/discussions/5](https://huggingface.co/nomic-ai/nomic-embed-text-v1.5-GGUF/discussions/5)
